android/greengrass-android-component-base/component-library/src/main/java/com/aws/greengrass/android/component/core/GreengrassComponentService.java

```diff
@@ -0,0 +1,876 @@
+     * request.
+     */
+    protected void onExitRequested() {
+        tryStop(EXIT_CODE_TERMINATED);
+    }
+
+    /**
+     * Service Start event callback.
+     * This method is invoked after the service has completed all the initialization, right before
+     * worker thread is started. Default implementation starts periodic ping of lifecycle observers.
+     */
+    protected void onServiceStart() {
+        PingThread thread = new PingThread();
+        thread.start();
+        pingThread.set(thread);
+    }
+
+    /**
+     * Service Stopped event callback.
+     * This method is invoked after the service work is finished, right before de-initialization is
+     * started. Default implementation does nothing.
+     */
+    protected void onServiceStop() {
+    }
+
+    /**
+     * No Lifecycle Observers Left event callback.
+     * This method is invoked when all the ping requests timed out. Default implementation stops
+     * the component.
+     */
+    protected void onNoObserversLeft() {
+        Log.w(LOG_TAG, "No observers left, terminating component");
+        if (componentWorkerThread != null && componentWorkerThread.isAlive()) {
+            componentWorkerThread.stopWorker();
+            return;
+        }
+    }
+
+    private void tryStop(int exitCode) {
+        tryStop(exitCode, true);
+    }
+
+    private void tryStop(int exitCode, boolean shouldJoin) {
+        if (!ensureStoppable()) {
+            return;
+        }
+
+        Log.d(LOG_TAG, "stop() exitCode " + exitCode + " shouldJoin " + shouldJoin);
+
+        if (componentWorkerThread != null && componentWorkerThread.isAlive()) {
+            componentWorkerThread.stopWorker();
+        }
+
+        BroadcastManager.sendBroadcast(this,
+                getPackageName(),
+                getMasterPackage(),
+                ACTION_COMPONENT_STOPPED,
+                exitCode);
+
+        Thread thread = workThread.getAndSet(null);
+        terminateThread(thread, shouldJoin);
+
+        for (Iterator<Messenger> it = lifecycleObservers.iterator(); it.hasNext(); ) {
+            Messenger messenger = it.next();
+            try {
+                messenger.send(Message.obtain(null, LIFECYCLE_MSG_EXIT_CODE, exitCode, 0));
+            } catch (RemoteException | IllegalStateException e) {
+                // probably an observer has crashed or terminated, remove from the set
+                it.remove();
+                Log.w(LOG_TAG, "Unable to notify remote node " + Log.getStackTraceString(e));
+            }
+        }
+
+        stopForeground(true);
+        stopSelf();
+        setStopped();
+    }
+
+    private void terminateThread(Thread thread, boolean join) {
+        Log.d(LOG_TAG, "terminateThread");
+        if (thread != null && thread.isAlive()) {
+            thread.interrupt();
+            if (join) {
+                try {
+                    thread.join(threadJoinTime.get());
+                } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)